controllers/binding/snapshotenvironmentbinding_controller.go

```diff
@@ -102,6 +110,12 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
 
 	adapter := NewAdapter(snapshotEnvironmentBinding, snapshot, environment, application, integrationTestScenario, logger, loader, r.Client, ctx)
 
+	bindingConditionsFailed := meta.FindStatusCondition(snapshotEnvironmentBinding.Status.BindingConditions, AppStudioSnapshotEnvironmentBindingConditionsFailed)
+	if bindingConditionsFailed != nil {
+		result, err := adapter.EnsureEphemeralEnvironmentsCleanedUp()
```

Comment:
The `EnsureEphemeralEnvironmentsCleanedUp` is part of the pipeline adapter and is specific to the reconciliation of Integration `pipelineRuns`. We would need to create a new function that handles this within the [snapshotenvironmentbinding adapter]([snapshotenvironmentbinding_adapter.go](https://github.com/redhat-appstudio/integration-service/blob/main/controllers/binding/snapshotenvironmentbinding_adapter.go)).

Reply:
Duplication of code is a good point. We could create a new function in [gitops/environment.go](https://github.com/redhat-appstudio/integration-service/blob/main/gitops/environment.go) which handles deletion of an environment in a generic manner, and that function could then be used by both adapter functions when needing to clean up an environment.